File: app/src/main/java/com/sdex/activityrunner/intent/IntentBuilderActivity.java
package com.sdex.activityrunner.intent;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.sdex.activityrunner.GetPremiumDialog;
import com.sdex.activityrunner.R;
import com.sdex.activityrunner.db.activity.ActivityModel;
import com.sdex.activityrunner.intent.LaunchParamsExtraListAdapter.Callback;
import com.sdex.activityrunner.intent.converter.LaunchParamsToIntentConverter;
import com.sdex.activityrunner.intent.dialog.ExtraInputDialog;
import com.sdex.activityrunner.intent.dialog.MultiSelectionDialog;
import com.sdex.activityrunner.intent.dialog.SingleSelectionDialog;
import com.sdex.activityrunner.intent.dialog.ValueInputDialog;
import com.sdex.activityrunner.intent.dialog.source.ActionSource;
import com.sdex.activityrunner.intent.dialog.source.CategoriesSource;
import com.sdex.activityrunner.intent.dialog.source.FlagsSource;
import com.sdex.activityrunner.intent.dialog.source.MimeTypeSource;
import com.sdex.activityrunner.intent.dialog.source.SelectionDialogSource;
import com.sdex.activityrunner.intent.history.HistoryActivity;
import com.sdex.activityrunner.util.IntentUtils;
import com.sdex.commons.BaseActivity;
import com.sdex.commons.ads.AdsDelegate;
import com.sdex.commons.ads.AppPreferences;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class IntentBuilderActivity extends BaseActivity
  implements ValueInputDialog.OnValueInputDialogCallback,
  SingleSelectionDialog.OnItemSelectedCallback,
  MultiSelectionDialog.OnItemsSelectedCallback,
  ExtraInputDialog.OnKeyValueInputDialogCallback {

  private static final String ARG_ACTIVITY_MODEL = "arg_activity_model";
  private static final String STATE_LAUNCH_PARAMS = "state_launch_params";

  private static final int EXTRAS_LIMIT = 3;

  @BindView(R.id.image_package_name)
  ImageView packageNameImageView;
  @BindView(R.id.package_name)
  TextView packageNameView;
  @BindView(R.id.image_class_name)
  ImageView classNameImageView;
  @BindView(R.id.class_name)
  TextView classNameView;
  @BindView(R.id.image_data)
  ImageView dataImageView;
  @BindView(R.id.data)
  TextView dataView;
  @BindView(R.id.image_action)
  ImageView actionImageView;
  @BindView(R.id.action)
  TextView actionView;
  @BindView(R.id.image_mime_type)
  ImageView mimeTypeImageView;
  @BindView(R.id.mime_type)
  TextView mimeTypeView;
  @BindView(R.id.image_extras)
  ImageView extrasImageView;
  @BindView(R.id.image_categories)
  ImageView categoriesImageView;
  @BindView(R.id.image_flags)
  ImageView flagsImageView;
  @BindView(R.id.add_extra)
  TextView addExtraView;
  @BindView(R.id.list_extras)
  RecyclerView listExtrasView;
  @BindView(R.id.list_categories)
  RecyclerView listCategoriesView;
  @BindView(R.id.list_flags)
  RecyclerView listFlagsView;

  private LaunchParams launchParams = new LaunchParams();

  private LaunchParamsListAdapter categoriesAdapter;
  private LaunchParamsListAdapter flagsAdapter;
  private LaunchParamsExtraListAdapter extraAdapter;
  private LaunchParamsViewModel viewModel;

  private AppPreferences appPreferences;
  private AdsDelegate adsDelegate;

  public static void start(Context context, ActivityModel activityModel) {
    Intent starter = new Intent(context, IntentBuilderActivity.class);
    starter.putExtra(ARG_ACTIVITY_MODEL, activityModel);
    context.startActivity(starter);
  }

  @Override
  protected int getLayout() {
    return R.layout.activity_intent_builder;
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    ButterKnife.bind(this);

    viewModel = ViewModelProviders.of(this).get(LaunchParamsViewModel.class);

    appPreferences = new AppPreferences(this);

    FrameLayout adsContainer = findViewById(R.id.ads_container);
    adsDelegate = new AdsDelegate(appPreferences, adsContainer);
    adsDelegate.initBanner(this, R.string.ad_banner_unit_id);

    enableBackButton();
    final ActivityModel activityModel = (ActivityModel)
      getIntent().getSerializableExtra(ARG_ACTIVITY_MODEL);
    if (activityModel != null) {
      setTitle(activityModel.getName());
      launchParams.setPackageName(activityModel.getPackageName());
      launchParams.setClassName(activityModel.getClassName());
    }

    if (savedInstanceState != null) {
      launchParams = savedInstanceState.getParcelable(STATE_LAUNCH_PARAMS);
    }

    configureRecyclerView(listExtrasView);
    configureRecyclerView(listCategoriesView);
    configureRecyclerView(listFlagsView);

    extraAdapter = new LaunchParamsExtraListAdapter(new Callback() {
      @Override
      public void onItemSelected(int position) {
        final LaunchParamsExtra extra = launchParams.getExtras().get(position);
        DialogFragment dialog = ExtraInputDialog.newInstance(extra, position);
        dialog.show(getSupportFragmentManager(), ExtraInputDialog.TAG);
      }

      @Override
      public void removeItem(int position) {
        launchParams.getExtras().remove(position);
        extraAdapter.notifyDataSetChanged();
        listExtrasView.requestLayout();
        updateExtrasAdd();
      }
    });
    extraAdapter.setHasStableIds(true);
    listExtrasView.setAdapter(extraAdapter);
    categoriesAdapter = new LaunchParamsListAdapter();
    categoriesAdapter.setHasStableIds(true);
    listCategoriesView.setAdapter(categoriesAdapter);
    flagsAdapter = new LaunchParamsListAdapter();
    flagsAdapter.setHasStableIds(true);
    listFlagsView.setAdapter(flagsAdapter);

    bindInputValueDialog(R.id.container_package_name, R.string.launch_param_package_name);
    bindInputValueDialog(R.id.container_class_name, R.string.launch_param_class_name);
    bindInputValueDialog(R.id.container_data, R.string.launch_param_data);
    bindSingleSelectionDialog(R.id.container_action, R.string.launch_param_action,
      new ActionSource());
    bindSingleSelectionDialog(R.id.container_mime_type, R.string.launch_param_mime_type,
      new MimeTypeSource());
    bindKeyValueDialog(R.id.container_extras);
    bindMultiSelectionDialog(R.id.categories_click_interceptor, R.string.launch_param_categories,
      new CategoriesSource());
    bindMultiSelectionDialog(R.id.flags_click_interceptor, R.string.launch_param_flags,
      new FlagsSource());

    findViewById(R.id.launch).setOnClickListener(v -> {
      viewModel.addToHistory(launchParams);
      LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
      final Intent intent = converter.convert();
      IntentUtils.launchActivity(IntentBuilderActivity.this, intent);
    });

    showLaunchParams();
  }

  @Override
  protected void onResume() {
    super.onResume();
    adsDelegate.detachBottomBannerIfNeed();
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putParcelable(STATE_LAUNCH_PARAMS, launchParams);
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    if (requestCode == HistoryActivity.REQUEST_CODE && resultCode == RESULT_OK) {
      launchParams = data.getParcelableExtra(HistoryActivity.RESULT);
      showLaunchParams();
    }
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.launch_param, menu);
    return super.onCreateOptionsMenu(menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.action_history: {
        final Intent intent = HistoryActivity.getLaunchIntent(this);
        startActivityForResult(intent, HistoryActivity.REQUEST_CODE);
        return true;
      }
      default:
        return super.onOptionsItemSelected(item);
    }
  }

  @Override
  public void onValueSet(int type, String value) {
    switch (type) {
      case R.string.launch_param_package_name:
        launchParams.setPackageName(value);
        break;
      case R.string.launch_param_class_name:
        launchParams.setClassName(value);
        break;
      case R.string.launch_param_data:
        launchParams.setData(value);
        break;
    }
    showLaunchParams();
  }

  @Override
  public void onItemSelected(int type, int position) {
    switch (type) {
      case R.string.launch_param_action:
        launchParams.setAction(position);
        break;
      case R.string.launch_param_mime_type:
        launchParams.setMimeType(position);
        break;
    }
    showLaunchParams();
  }

  @Override
  public void onItemsSelected(int type, ArrayList<Integer> positions) {
    switch (type) {
      case R.string.launch_param_categories:
        launchParams.setCategories(positions);
        break;
      case R.string.launch_param_flags:
        launchParams.setFlags(positions);
        break;
    }
    showLaunchParams();
  }

  @Override
  public void onValueSet(LaunchParamsExtra extra, int position) {
    final ArrayList<LaunchParamsExtra> extras = launchParams.getExtras();
    if (position == -1) {
      extras.add(extra);
    } else {
      extras.set(position, extra);
    }
    showLaunchParams();
  }

  private void configureRecyclerView(RecyclerView recyclerView) {
    recyclerView.setNestedScrollingEnabled(false);
    recyclerView.setHasFixedSize(true);
  }

  private void bindInputValueDialog(int viewId, int type) {
    findViewById(viewId).setOnClickListener(v -> {
      String initialValue = getValueInitialPosition(type);
      ValueInputDialog dialog = ValueInputDialog.newInstance(type, initialValue);
      dialog.show(getSupportFragmentManager(), ValueInputDialog.TAG);
    });
  }

  private void bindSingleSelectionDialog(int viewId, int type, SelectionDialogSource source) {
    findViewById(viewId).setOnClickListener(v -> {
      int initialPosition = getSingleSelectionInitialPosition(type);
      SingleSelectionDialog dialog =
        SingleSelectionDialog.newInstance(type, source, initialPosition);
      dialog.show(getSupportFragmentManager(), SingleSelectionDialog.TAG);
    });
  }

  private void bindMultiSelectionDialog(int viewId, int type, SelectionDialogSource source) {
    findViewById(viewId).setOnClickListener(v -> {
      ArrayList<Integer> initialPositions = getMultiSelectionInitialPositions(type);
      MultiSelectionDialog dialog =
        MultiSelectionDialog.newInstance(type, source, initialPositions);
      dialog.show(getSupportFragmentManager(), MultiSelectionDialog.TAG);
    });
  }

  private void bindKeyValueDialog(int viewId) {
    findViewById(viewId).setOnClickListener(v -> {
      final int size = launchParams.getExtras().size();
      if (size >= EXTRAS_LIMIT && !appPreferences.isProVersion()) {
        GetPremiumDialog dialog =
          GetPremiumDialog.newInstance(R.string.pro_version_unlock_extras);
        dialog.show(getSupportFragmentManager(), GetPremiumDialog.TAG);
        return;
      }
      DialogFragment dialog = ExtraInputDialog.newInstance(null, -1);
      dialog.show(getSupportFragmentManager(), ExtraInputDialog.TAG);
    });
  }

  private String getValueInitialPosition(int type) {
    switch (type) {
      case R.string.launch_param_package_name:
        return launchParams.getPackageName();
      case R.string.launch_param_class_name:
        return launchParams.getClassName();
      case R.string.launch_param_data:
        return launchParams.getData();
      default:
        throw new IllegalStateException("Unknown type " + type);
    }
  }

  private int getSingleSelectionInitialPosition(int type) {
    switch (type) {
      case R.string.launch_param_action:
        return launchParams.getAction();
      case R.string.launch_param_mime_type:
        return launchParams.getMimeType();
      default:
        throw new IllegalStateException("Unknown type " + type);
    }
  }

  private ArrayList<Integer> getMultiSelectionInitialPositions(int type) {
    switch (type) {
      case R.string.launch_param_categories:
        return launchParams.getCategories();
      case R.string.launch_param_flags:
        return launchParams.getFlags();
      default:
        throw new IllegalStateException("Unknown type " + type);
    }
  }

  private void showLaunchParams() {
    final String packageName = launchParams.getPackageName();
    packageNameView.setText(packageName);
    updateIcon(packageNameImageView, packageName);
    final String className = launchParams.getClassName();
    classNameView.setText(className);
    updateIcon(classNameImageView, className);
    final String data = launchParams.getData();
    dataView.setText(data);
    updateIcon(dataImageView, data);
    final String actionValue = launchParams.getActionValue();
    actionView.setText(actionValue);
    updateIcon(actionImageView, actionValue);
    final String mimeTypeValue = launchParams.getMimeTypeValue();
    mimeTypeView.setText(mimeTypeValue);
    updateIcon(mimeTypeImageView, mimeTypeValue);
    final ArrayList<LaunchParamsExtra> extras = launchParams.getExtras();
    extraAdapter.setItems(extras);
    updateIcon(extrasImageView, extras);
    updateExtrasAdd();
    final ArrayList<String> categoriesValues = launchParams.getCategoriesValues();
    categoriesAdapter.setItems(categoriesValues);
    updateIcon(categoriesImageView, categoriesValues);
    final ArrayList<String> flagsValues = launchParams.getFlagsValues();
    flagsAdapter.setItems(flagsValues);
    updateIcon(flagsImageView, flagsValues);
  }

  private void updateExtrasAdd() {
    final ArrayList<LaunchParamsExtra> extras = launchParams.getExtras();
    if (extras.isEmpty()) {
      addExtraView.setVisibility(View.GONE);
    } else {
      addExtraView.setVisibility(View.VISIBLE);
    }
  }

  private void updateIcon(ImageView imageView, String text) {
    imageView.setImageResource(TextUtils.isEmpty(text) ?
      R.drawable.ic_assignment : R.drawable.ic_assignment_done);
  }

  private void updateIcon(ImageView imageView, List list) {
    imageView.setImageResource((list == null || list.isEmpty()) ?
      R.drawable.ic_assignment : R.drawable.ic_assignment_done);
  }
}

File: app/src/main/java/com/sdex/activityrunner/intent/IntentBuilderActivity.kt
package com.sdex.activityrunner.intent

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.annotation.StringRes
import androidx.core.view.isVisible
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.sdex.activityrunner.R
import com.sdex.activityrunner.app.ActivityModel
import com.sdex.activityrunner.commons.BaseActivity
import com.sdex.activityrunner.databinding.ActivityIntentBuilderBinding
import com.sdex.activityrunner.extensions.parcelable
import com.sdex.activityrunner.extensions.serializable
import com.sdex.activityrunner.intent.LaunchParamsExtraListAdapter.Callback
import com.sdex.activityrunner.intent.converter.LaunchParamsToShellCommandConverter
import com.sdex.activityrunner.intent.dialog.ExtraInputDialog
import com.sdex.activityrunner.intent.dialog.MultiSelectionDialog
import com.sdex.activityrunner.intent.dialog.SingleSelectionDialog
import com.sdex.activityrunner.intent.dialog.ValueInputDialog
import com.sdex.activityrunner.intent.history.HistoryActivity
import com.sdex.activityrunner.preferences.AppPreferences
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class IntentBuilderActivity : BaseActivity() {

    @Inject
    lateinit var appPreferences: AppPreferences

    private val viewModel: LaunchParamsViewModel by viewModels()
    private lateinit var binding: ActivityIntentBuilderBinding

    private val categoriesAdapter = LaunchParamsListAdapter()
    private val flagsAdapter = LaunchParamsListAdapter()
    private val extraAdapter = LaunchParamsExtraListAdapter()

    private val pickHistoryItem =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            it.data?.let { intent ->
                val result = intent.parcelable<LaunchParams>(HistoryActivity.RESULT)
                viewModel.setLaunchParams(result)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityIntentBuilderBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupToolbar(isBackButtonEnabled = true)

        val activityModel = intent.serializable<ActivityModel>(ARG_ACTIVITY_MODEL)
        viewModel.initialize(activityModel)

        title = activityModel?.name ?: getString(R.string.intent_launcher_activity)

        binding.listExtrasView.configureRecyclerView()
        binding.listCategoriesView.configureRecyclerView()
        binding.listFlagsView.configureRecyclerView()

        extraAdapter.callback = object : Callback {
            override fun onItemSelected(position: Int) {
                val extra = viewModel.launchParamsState.value.extras[position]
                val dialog = ExtraInputDialog.newInstance(extra, position)
                dialog.show(supportFragmentManager, ExtraInputDialog.TAG)
            }

            @SuppressLint("NotifyDataSetChanged")
            override fun removeItem(position: Int) {
                viewModel.removeExtra(position)
                extraAdapter.notifyDataSetChanged()
                binding.listExtrasView.requestLayout()
            }
        }
        extraAdapter.setHasStableIds(true)
        binding.listExtrasView.adapter = extraAdapter

        categoriesAdapter.setHasStableIds(true)
        binding.listCategoriesView.adapter = categoriesAdapter

        flagsAdapter.setHasStableIds(true)
        binding.listFlagsView.adapter = flagsAdapter

        bindInputValueDialog(binding.containerPackageName, R.string.launch_param_package_name)
        bindInputValueDialog(binding.containerClassName, R.string.launch_param_class_name)
        bindInputValueDialog(binding.containerData, R.string.launch_param_data)
        bindInputValueDialog(binding.actionEditImageView, R.string.launch_param_action)
        bindInputValueDialog(binding.mimeTypeEditImageView, R.string.launch_param_mime_type)
        bindSingleSelectionDialog(binding.containerAction, R.string.launch_param_action)
        bindSingleSelectionDialog(binding.containerMimeType, R.string.launch_param_mime_type)
        bindKeyValueDialog(binding.containerExtras)
        bindMultiSelectionDialog(
            binding.categoriesClickInterceptor,
            R.string.launch_param_categories,
        )
        bindMultiSelectionDialog(binding.flagsClickInterceptor, R.string.launch_param_flags)

        binding.launchWithRoot.setOnCheckedChangeListener { _, isChecked ->
            viewModel.setUseRoot(isChecked)
        }

        binding.launchWithRoot.isChecked = (activityModel?.launchRequiresRoot == true)

        binding.launch.setOnClickListener {
            viewModel.launch(binding.saveToHistory.isChecked)
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.launchParamsState.collect {
                        showLaunchParams(it)
                    }
                }
                launch {
                    viewModel.events.collect {
                        handleLaunchEvent(it)
                    }
                }
            }
        }
    }

    private fun handleLaunchEvent(event: LaunchParamsViewModel.LaunchEvent) {
        when (event) {
            is LaunchParamsViewModel.LaunchEvent.LaunchError ->
                showErrorDialog(R.string.starting_activity_intent_failed, event.details)

            LaunchParamsViewModel.LaunchEvent.RootUnavailable ->
                Toast.makeText(this, R.string.starting_activity_root_not_available, Toast.LENGTH_SHORT)
                    .show()

            LaunchParamsViewModel.LaunchEvent.RootSuccess -> if (appPreferences.isShowLaunchToast) {
                Toast.makeText(this, R.string.starting_activity_intent, Toast.LENGTH_SHORT).show()
            }

            is LaunchParamsViewModel.LaunchEvent.RootError ->
                showErrorDialog(R.string.starting_activity_root_error, event.details)
        }
    }

    private fun showErrorDialog(@StringRes titleRes: Int, message: String?) {
        MaterialAlertDialogBuilder(this)
            .setTitle(titleRes)
            .setMessage(message)
            .setPositiveButton(android.R.string.ok, null)
            .show().apply {
                val messageTextView = findViewById<TextView>(android.R.id.message)
                messageTextView?.setTextIsSelectable(true)
            }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.launch_param, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_history -> {
                pickHistoryItem.launch(HistoryActivity.getLaunchIntent(this))
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun RecyclerView.configureRecyclerView() {
        isNestedScrollingEnabled = false
        setHasFixedSize(false)
    }

    private fun bindInputValueDialog(view: View, type: Int) {
        view.setOnClickListener {
            val initialValue = viewModel.getValueInitialValue(type)
            val dialog = ValueInputDialog.newInstance(type, initialValue)
            dialog.show(supportFragmentManager, ValueInputDialog.TAG)
        }
    }

    private fun bindSingleSelectionDialog(view: View, type: Int) {
        view.setOnClickListener {
            val initialPosition = viewModel.getSingleSelectionInitialPosition(type)
            val dialog = SingleSelectionDialog.newInstance(type, initialPosition)
            dialog.show(supportFragmentManager, SingleSelectionDialog.TAG)
        }
    }

    private fun bindMultiSelectionDialog(view: View, type: Int) {
        view.setOnClickListener {
            val initialPositions = viewModel.getMultiSelectionInitialPositions(type)
            val dialog = MultiSelectionDialog.newInstance(type, initialPositions)
            dialog.show(supportFragmentManager, MultiSelectionDialog.TAG)
        }
    }

    private fun bindKeyValueDialog(view: View) {
        view.setOnClickListener {
            val dialog = ExtraInputDialog.newInstance(null, -1)
            dialog.show(supportFragmentManager, ExtraInputDialog.TAG)
        }
    }

    private fun showLaunchParams(launchParams: LaunchParams) {
        binding.packageNameView.text = launchParams.packageName
        binding.classNameView.text = launchParams.className
        binding.dataView.text = launchParams.data
        binding.actionView.text = launchParams.action
        binding.mimeTypeView.text = launchParams.mimeType
        extraAdapter.setItems(launchParams.extras)
        categoriesAdapter.setItems(launchParams.getCategoriesValues())
        flagsAdapter.setItems(launchParams.getFlagsValues())
        if (binding.launchWithRoot.isChecked != launchParams.useRoot) {
            binding.launchWithRoot.isChecked = launchParams.useRoot
        }
        binding.rootExtrasWarning.isVisible = launchParams.useRoot &&
            LaunchParamsToShellCommandConverter.getUnsupportedExtras(launchParams).isNotEmpty()
        updateExtrasAdd(launchParams)
    }

    private fun updateExtrasAdd(launchParams: LaunchParams) {
        binding.addExtraView.isVisible = launchParams.extras.isNotEmpty()
    }

    companion object {

        private const val ARG_ACTIVITY_MODEL = "arg_activity_model"

        fun start(context: Context, model: ActivityModel?) {
            val starter = Intent(context, IntentBuilderActivity::class.java)
            starter.putExtra(ARG_ACTIVITY_MODEL, model)
            context.startActivity(starter)
        }
    }
}
